security/security/src/main/java/io/helidon/security/Grant.java

```diff
@@ -59,9 +59,11 @@ protected Grant(Builder<?> builder) {
      * Creates a fluent API builder to build new instances of this class.
      *
      * @return builder instance
+     * @param <T> type of builder
      */
-    public static Builder<?> builder() {
-        return new Builder<>();
+    @SuppressWarnings("unchecked")
+    public static <T extends Builder<T>> T builder() {
+        return (T) new Builder<T>();
```

Comment:
This cannot be done, as the `builder` method always returns the same instance, which may not fit into the return type

Reply:
We should change this to have:
`AbstractBuilder` class that has type definition (`AbstractBuilder<B extends AbstractBuilder<B>`) and the `Builder` class that `extends AbstractBuilder<Builder>` and this method should return `Builder` explicitly.
Then other builders extends `AbstractBuilder` as well.

Not sure if this could be done in a backward compatible way here

Reply:
We should change this to have:
`AbstractBuilder` class that has type definition (`AbstractBuilder<B extends AbstractBuilder<B>`) and the `Builder` class that `extends AbstractBuilder<Builder>` and this method should return `Builder` explicitly.
Then other builders extends `AbstractBuilder` as well.

Not sure if this could be done in a backward compatible way here

Reply:
> This cannot be done, as the `builder` method always returns the same instance, which may not fit into the return type

You mean the implementation was incorrect before?

Reply:
> We should change this to have: `AbstractBuilder` class that has type definition (`AbstractBuilder<B extends AbstractBuilder<B>`) and the `Builder` class that `extends AbstractBuilder<Builder>` and this method should return `Builder` explicitly. Then other builders extends `AbstractBuilder` as well.
> 
> Not sure if this could be done in a backward compatible way here

Please see the discussion in the issue. The static `builder` method in a subclass cannot return a type that isn't related to that returned in the base class. So if Grant.Builder <: AbstractBuilder and Role.Builder <: AbstractBuilder, then the `builder` method cannot return the corresponding subtypes.